modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,359 @@
+	}
+}
+
+// fileOffsetFromRootIndex calculates the offset of the merkle root at index i from
+// the beginning of the contract file.
+func fileOffsetFromRootIndex(i int) int64 {
+	return contractHeaderSize + crypto.HashSize*int64(i)
+}
+
+// delete deletes the sector root at a certain index.
+func (mr *merkleRoots) delete(i int) error {
+	// Check if the index is correct
+	if i >= mr.numMerkleRoots {
+		build.Critical("can't delete non-existing root")
+		return nil
+	}
+	// If i is the index of the last element we call deleteLastRoot.
+	if i == mr.numMerkleRoots-1 {
+		return mr.deleteLastRoot()
+	}
+	// If we don't have any uncached roots we need to delete the last cached
+	// tree and add its elements to the uncached roots.
+	if len(mr.uncachedRoots) == 0 {
+		mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+		rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+		roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+		if err != nil {
+			return errors.AddContext(err, "failed to read cached tree's roots")
+		}
+		mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	}
+	// Swap the root at index i with the last root in mr.uncachedRoots.
+	_, err := mr.file.WriteAt(mr.uncachedRoots[len(mr.uncachedRoots)-1][:], fileOffsetFromRootIndex(i))
+	if err != nil {
+		return errors.AddContext(err, "failed to swap root to delete with last one")
+	}
+	// If the deleted root was not cached we swap the roots in memory too.
+	// Otherwise we rebuild the cachedSubTree.
+	if index, cached := mr.isIndexCached(i); !cached {
+		mr.uncachedRoots[index] = mr.uncachedRoots[len(mr.uncachedRoots)-1]
+	} else {
+		err = mr.rebuildCachedTree(index)
+	}
+	if err != nil {
+		return errors.AddContext(err, "failed to rebuild cached tree")
+	}
+	// Now that the element we want to delete is the last root we can simply
+	// delete it by calling mr.deleteLastRoot.
+	return mr.deleteLastRoot()
+}
+
+// deleteLastRoot deletes the last sector root of the contract.
+func (mr *merkleRoots) deleteLastRoot() error {
+	// Decrease the numMerkleRoots counter.
+	mr.numMerkleRoots--
+	// Truncate file to avoid interpreting trailing data as valid.
+	if err := mr.file.Truncate(fileOffsetFromRootIndex(mr.numMerkleRoots)); err != nil {
+		return errors.AddContext(err, "failed to delete last root from file")
+	}
+	// If the last element is uncached we can simply remove it from the slice.
+	if len(mr.uncachedRoots) > 0 {
+		mr.uncachedRoots = mr.uncachedRoots[:len(mr.uncachedRoots)-1]
+		return nil
+	}
+	// If it is not uncached we need to delete the last cached tree and load
+	// its elements into mr.uncachedRoots.
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
```

Comment:
this logic of deleting the last cached and populating the uncached appears in a few places, probably worth combining them into a helper method.